bigtable-hbase-parent/bigtable-hbase/src/main/java/org/apache/hadoop/hbase/client/AbstractBigtableAdmin.java

```diff
@@ -819,93 +832,174 @@ public boolean isTableAvailable(TableName tableName, byte[][] splitKeys) throws
   @Override
   public void snapshot(String snapshotName, TableName tableName)
       throws IOException, SnapshotCreationException, IllegalArgumentException {
-    throw new UnsupportedOperationException("snapshot");  // TODO
+    waitForOperation(snapshotTable(snapshotName, tableName));
+  }
+
+  private Operation snapshotTable(String snapshotName, TableName tableName)
+      throws IOException {
+    return bigtableTableAdminClient.snapshotTable(SnapshotTableRequest.newBuilder()
+        .setCluster(getClusterName().toString())
+        .setSnapshotId(snapshotName)
+        .setName(options.getInstanceName().toTableNameStr(tableName.getNameAsString()).toString())
+        .build());
+  }
+
+  /**
+   * Waits for an operation like creating a snapshot to complete.
+   * @param operation. The current state of the operation.
+   * @throws InterruptedException if a user interrupts the process, usually with a ^C.
+   * @throws IOException
+   */
+  private void waitForOperation(Operation operation) throws IOException {
+    GetOperationRequest request =
+        GetOperationRequest.newBuilder().setName(operation.getName()).build();
+    Operation currentOperationState = operation;
+    while (true) {
+      if (currentOperationState.getDone()) {
+        switch (currentOperationState.getResultCase()) {
+        case RESPONSE:
+          return;
+        case ERROR:
+          throw new RuntimeException(
+              "Cluster could not be resized: " + currentOperationState.getError());
+        case RESULT_NOT_SET:
+          throw new IllegalStateException(
+              "System returned invalid response for Operation check: " + currentOperationState);
+        }
+      }
+      try {
+        Thread.sleep(200);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException("Wating for operation was interrupted.", e);
+      }
+      currentOperationState =
```

Comment:
5 seconds per logging seems reasonable.  How long should we wait before we timeout?